src/Buffer.cpp

```diff
@@ -3,16 +3,32 @@
 #include "Error.h"
 #include "JITModule.h"
 #include "runtime/HalideRuntime.h"
+#include "Target.h"
 
 namespace Halide {
 namespace Internal {
 
 namespace {
-void check_buffer_size(uint64_t bytes, const std::string &name) {
-    user_assert(bytes < (1UL << 31)) << "Total size of buffer " << name << " exceeds 2^31 - 1\n";
-}
+
+uint64_t multiply_buffer_size_check_overflow(uint64_t size, uint64_t factor, const std::string &name) {
+    // Ignore the dimensions for which the extent is zero.
+    if (!factor) return size;
+
+    // Multiply and check for 64-bit overflow
+    uint64_t result = size * factor;
+    bool overflow = (result / factor) != size;
+
+    // Check against the limits Halide internally assumes in its compiled code.
+    overflow |= (sizeof(void *) == 4) && (result >= (1UL << 31));
```

Comment:
This seems to be automatically choosing large buffers if compiled in 64-bit mode. Is this always true? I would expect in the default case for both 32-bit and 64-bit processes that LargeBuffers is not set. Two possible solutions are to get the JIT target to decide which check to use, or to change the comment to say this is a conservative check and that in a 64-bit process a filter is not using the LargeBuffer feature may still signal a buffer too large error at call time.

Also I might use sizeof(size_t) rather than sizeof(void *). Not 100% sure, but in theory 64-bit pointer with 32-bit max allocation is a possible thing.